File: headless/level.py
import os
from typing import Dict, List
import logging
import random
import itertools

import consts
import utils
import zombie
import plant


class Level():
    """
    This is the Env, generated each episode.
    A Level starts from scratch (as in the regular game).
    A Level gets it's zomble spawn times at __init__, from json or at random (this will be a flag for Level.__init__())

    As a general rule, the 2D grids unpack in the following way:
    grid[height][width]. For example:
    grid[3][2] means the square that's fourth from the top and third from the left.
    grid[0] means the top row.
    
    Note: Right now, it's the responsibility of the attacker to remove the object it attacked from all data structures if needed.

    TODO: Carefully consider the right Data structures to use for efficient management of objects in the environment. Check out dicts - O(1) removal
    """
    def __init__(self, columns, lanes, level_data: dict, chosen_plants: List[str], fps = 30, logfile=consts.LOG_FILE_NAME):
        # Technical data:
        self.fps = fps
        self.frame = 0
        self.lanes = lanes # type: int
        self.columns = columns # type: int
        self.done = False
        self.win = False
        self.suns = 50 # Bank value
        self.chosen_plants = chosen_plants
        self.plant_available_frame = {plant_name: 0 for plant_name in chosen_plants}
        self.plant_costs = utils.get_plant_costs()
        
        # Object data
        self.replant_queue = {plant_name: 0 for plant_name in utils.get_plant_names()}
        self.zombies = [] # type: list[zombie.Zombie]
        self.plants = [] # type: list[plant.Plant]
        self.active_suns = []
        self.bullets = [] # type: list[plant.Bullet]
        self.home_column = [True] * lanes # type: list[str] # This is a special column that contains either a lawnmower, nothing or a zombie
        self.zombie_grid = [[[] for _ in range(columns)] for _ in range(lanes)] # type: list[list[list[zombie.Zombie]]]
        self.plant_grid = [[None for _ in range(columns)] for _ in range(lanes)] # type: list[list[plant.Plant]]
        
        # Internal data
        self.sun_value = 25
        self.last_sun_generated_frame = 0
        self.sun_interval = 10
        self.level_data = level_data # type: dict
        self.zombies_to_be_spawned = utils.get_zombies_to_be_spawned(level_data)

        utils.configure_logging(logfile)
        
    # def assign_zombie_damage(self):
    #     for zombie in self.zombies:
    #         zombie.attack(self)

    # def assign_plant_damage(self):
    #     for plant in self.plants: # All plants try to shoot or attack
    #         plant.do_action(self)
        
    #     # TODO - [note] We're copying the bullets array here because we're removing objects from it while iterating
    #     for bullet in self.bullets[:]: # All bullets either hit a target or move. New bullet can hit target on same frame as it's created
    #         bullet.attack_or_move(self)

    # def move_zombies(self):
    #     for zombie in self.zombies:
    #         zombie.move(self)

    def do_plant_actions(self):
        for plant in self.plants[:]:
            plant.do_action(self)
        
        for bullet in self.bullets[:]:
            bullet.attack_or_move(self)

    def do_zombie_actions(self):
        for zombie in self.zombies[:]:
            zombie.do_action(self)

    def spawn_zombies(self):
        """
        level_data should be a dict of the following format:
        "frame number": [[zombie_type, zombie_lane], [zombie_type, zombie_lane]]
        for example:
        600: [[normal, 0], [conehead, 3], [buckethead, 5]]
        """
        if not self.zombies_to_be_spawned:
            return
        
        curr_sec = str(self.frame // self.fps)
        if self.frame % self.fps == 0 and self.zombies_to_be_spawned.get(curr_sec):
            for zombie_type, lane in self.zombies_to_be_spawned[curr_sec]:
                logging.debug(f"[{self.frame}] {zombie_type} Zombie being spawned at lane {lane}")
                new_zombie = zombie.Zombie(zombie_type)
                new_zombie.lane = lane
                new_zombie.column = self.columns - 1
                new_zombie.last_moved = self.frame
                # self.zombies[new_zombie] = new_zombie.pos
                self.zombies.append(new_zombie)
                self.zombie_grid[lane][self.columns - 1].append(new_zombie)
            self.zombies_to_be_spawned.pop(curr_sec)

    def spawn_suns(self):
        if (self.frame - self.last_sun_generated_frame) > self.sun_interval * self.fps:
            self.last_sun_generated_frame = self.frame
            if consts.AUTO_COLLECT:
                self.suns += 25
            else:
                self.active_suns.append([0, 0]) # TODO: Randomize sun location
            logging.debug(f"[{self.frame}] Sun autospawned. Total: {self.suns}.")

    def _is_plant_legal(self, plant_name: str, lane, column):
        # Are the provided coords within the map?
        if lane < 0 or lane >= self.lanes or column < 0 or column >= self.columns:
            return False
        # Was this plant selected for this run?
        if plant_name not in self.chosen_plants:
            return False
        # Location is free
        if self.plant_grid[lane][column]:
            return False
        # There's no need to recharge
        if self.plant_available_frame[plant_name] > self.frame:
            return False
        # There's enough suns
        if self.suns < self.plant_costs[plant_name]:
            return False
        return True
        
    def action_is_legal(self, action): # This is for the player to check if he can do an action right now
        if not action:
            return True
        
        if action[0] == "plant":
            _, plant_name, lane, column = action
            return self._is_plant_legal(plant_name, lane, column)

        return False
        
    def plant(self, plant_name: str, lane, column):
        if not self._is_plant_legal(plant_name, lane, column):
            return
        # new_plant = plant.name_to_class[plant_name](lane, column) # type: plant.Plant
        new_plant = plant.create_plant_instance(plant_name, lane, column, self.frame) # selector for correct type of subclass, check if this can be done more cleanly
        self.plants.append(new_plant)
        self.plant_grid[lane][column] = new_plant
        self.suns -= new_plant.cost
        self.plant_available_frame[plant_name] = self.frame + (new_plant.recharge * self.fps) # Next frame the plant is plantable is this frame + cooldown frames
        logging.debug(f"[{self.frame}] Planted {plant_name} in {lane, column}. Total: {self.suns}.")
        
    def do_player_action(self, action: list):
        if not action:
            return
        if action[0] == "plant":
            _, plant_name, lane, column = action
            self.plant(plant_name, lane, column)
    
    def check_done(self):
        for lane in range(self.lanes):
            last_tile = self.zombie_grid[lane][0]
            if not last_tile: # if there are no zombies in the last tile of this lane
                continue
            for zombie in last_tile:
                if zombie.reached_house: # Zombie is about to enter the house
                    if self.home_column[lane]: # There is a lawnmower in this lane
                        self.home_column[lane] = False
                        logging.debug(f"[{self.frame}] Zombie in {zombie.lane, zombie.column} triggered a lawnmower.")
                        active_lawnmower = plant.Lawnmower(lane)
                        self.bullets.append(active_lawnmower)
                        active_lawnmower.attack(self)
                    else:
                        self.done = True
                        self.win = False
                        logging.debug(f"[{self.frame}] You've lost! Zombie entered at lane {zombie.lane}")
                        return

        if not self.done and not self.zombies_to_be_spawned and not self.zombies:
            # No zombies left to spawn or on the map
            self.done = True
            self.win = True
            logging.debug(f"[{self.frame}] You've won!")
            return
        
    def construct_state(self):
        grid = [[[] for _ in range(self.columns)] for _ in range(self.lanes)]
        for plant in self.plants:
            lane, column = plant.lane, plant.column
            grid[lane][column].append(plant.__repr__())
        for zombie in self.zombies:
            grid[zombie.lane][zombie.column].append(zombie.__repr__())
        state_tuple = (self.suns, self.home_column, grid)
        return state_tuple

    def sample_action(self):
        """
        This function returns a random legal action from the action space.
        If there are no legal action, returns None.
        """
        affordable_plants = []
        for plant in self.chosen_plants:
            if self.plant_costs[plant] <= self.suns and self.plant_available_frame[plant] < self.frame:
                affordable_plants.append(plant)
        if not affordable_plants:
            return None # No plants are affordable
        chosen_plant = random.choice(affordable_plants)
        open_positions = []
        for position in itertools.product(range(self.lanes), range(self.columns)):
            if not self.plant_grid[position[0]][position[1]]:
                open_positions.append(position)
        if not open_positions:
            return None # No positions are open
        chosen_position = random.choice(open_positions)
        return ["plant", chosen_plant, chosen_position[0], chosen_position[1]]

    def step(self, action: str):
        """
        The main function of the environment ("Level")
        Every time step() is called, (at least) the following must happen:
        1. Zombies assign damage and move
        2. Plants assign damage
        4. New zombs are generated (as needed)
        5. Suns are generated (as objects? straight into bank? Maybe leave this as a setting of the Level __init__)
        6. Player can use oprerators to interact with env
        Note: one step corresponds to one frame (in a 60 fps game). As such, things wont acutally happen at every step.
        For example, plants will attack every ~60-120 frames (depending on plant), suns are auto-generated every ~600 frames 
        """
        self.do_plant_actions()
        self.do_zombie_actions()
        self.spawn_zombies()
        self.spawn_suns()
        self.do_player_action(action)
        self.check_done()
        self.frame += 1
        return self.construct_state()

    def print_grid(self):
        os.system('clear')
        for row in range(self.lanes):
            print("-" * (4 * self.lanes))
            for col in range(self.columns):
                cell_content = ["0", "0", "0"]  # Plants, Bullets, Zombies
                if self.plant_grid[row][col]:
                    if type(self.plant_grid[row][col]) == plant.Peashooter:
                        cell_content[0] = "P"
                    if type(self.plant_grid[row][col]) == plant.Sunflower:
                        cell_content[0] = "S"
                if self.zombie_grid[row][col]:
                    cell_content[2] = len(self.zombie_grid[row][col])
                for bullet in self.bullets:
                    if bullet.position == (row, col):
                        cell_content[1] += 1
                print(f"| {cell_content[0]}{cell_content[1]}{cell_content[2]} ", end='')
            print("|")
        print("-" * (4 * self.columns))
                
                
                
    # def print_board(matrix):
    # for row in matrix:
    #     print("-" * (4 * len(row) + 1))  # Print horizontal line
    #     for element in row:
    #         print("|", f" {element} ", end='')  # Print element with vertical lines
    #     print("|")  # Print closing vertical line
    # print("-" * (4 * len(matrix[0]) + 1))  # Print bottom horizontal line

File: cpp_test/level.cpp
#include "level.h"
#include <assert.h>
#include <iostream>
#include <string>
#include <sstream>
#include <list>

#ifndef DNDEBUG
#define LOG_FRAME(frame, msg) std::cout << "[" << frame << "] " << msg << endl;
#else
#define LOG_FRAME(frame, msg)
#endif

using std::string;

using std::cout;
using std::endl;

Level::Level(int lanes, int columns, int fps){
    lanes = lanes;
    cols = cols;
    suns = 50;
    frame = 1;
    last_sun_generated = 1;
    zombie_in_home_col = false;
    done = false;
    win = false;
    lawnmowers = new bool[lanes];
    for (int i = 0; i < lanes; i++){
        lawnmowers[i] = true;
    }
    zombie_list = std::list<Zombie*>();
    zombie_grid = new std::list<Zombie*>*[lanes];
    for (int i = 0; i < lanes; i++){
        zombie_grid[i] = new std::list<Zombie*>[cols];
    }
    plant_list = std::list<Plant*>();
    plant_grid = new Plant**[lanes];
    for (int i = 0; i < lanes; i++){
        plant_grid[i] = new Plant*[cols];
        for (int j = 0; j < columns; j++){
            plant_grid[i][j] = nullptr;
        }
    }
    zombies_to_spawn = vector<Zombie2Spawn>();
    zombies_to_spawn.push_back(Zombie2Spawn(100, 2));
    zombies_to_spawn.push_back(Zombie2Spawn(300, 2));
}

bool Level::is_action_legal(string plant_name, int lane, int col){
    if (this->done == true){
        return false;
    }
    if (plant_name == "no_action"){
        return true;
    }
    if (lane >= this->lanes || col >= this->cols) {
        return false;
    }
    if (this->plant_grid[lane][col] != nullptr) {
        return false;
    }
    if (plant_name != "sunflower" && plant_name != "peashooter"){
        return false;
    }
    if (plant_name == "sunflower" && this->suns < 50){
        return false;
    }
    if (plant_name == "peashooter" && this->suns < 100){
        return false;
    }
}

void Level::plant(int lane, int column, bool isSunflower) {
    Plant* new_plant = new Plant(lane, column, this->frame, isSunflower);
    this->plant_list.push_front(new_plant);
    this->plant_grid[lane][column] = new_plant;
}
void Level::do_zombie_actions(){
    for (Zombie* zombie : this->zombie_list){
        zombie->do_action(this);
    }
}
void Level::do_plant_actions(){
    for (Plant* plant : this->plant_list){
        plant->do_action(this);
    }
}
void Level::do_player_action(string plant_name, int lane, int col){
    if (this->is_action_legal(plant_name, lane, col) == false){
        LOG_FRAME(this->frame, "ILLEGAL ACTION");
        return;
    }
    if (plant_name == "sunflower") {
        this->plant(lane, col, true);
        std::stringstream log_msg;
        log_msg << "planted sunflower at lane " << lane << " col " << col;
        LOG_FRAME(this->frame, log_msg.str());
    }
    else if (plant_name == "peashooter"){
        this->plant(lane, col, false);
        std::stringstream log_msg;
        log_msg << "planted peashooter at lane " << lane << " col " << col;
        LOG_FRAME(this->frame, log_msg.str());
    }
    else if (plant_name == "no_action"){
        // do nothing
        LOG_FRAME(this->frame, "no action");
    }
}
void Level::spawn_zombies(){
    while (!this->zombies_to_spawn.empty() && this->zombies_to_spawn.back().frame == this->frame) {
        Zombie2Spawn zombie_template = this->zombies_to_spawn.back();
        this->zombies_to_spawn.pop_back();
        Zombie* new_zombie = new Zombie(zombie_template.lane, this->cols - 1, this); // TODO - column
        this->zombie_list.push_back(new_zombie);
        this->zombie_grid[new_zombie->lane][new_zombie->col].push_back(new_zombie);
    }
}
void Level::spawn_suns(){
    if (this->frame - this->last_sun_generated >= 100) {
        assert(this->frame - this->last_sun_generated == 100);
        this->suns += 25;
        this->last_sun_generated = this->frame;
        LOG_FRAME(this->frame, "generated sun");
    }
}
void Level::check_endgame(){
    if (this->zombie_in_home_col == true){
        bool kill_lane = false;
        this->zombie_in_home_col = false;
        for (int lane = 0; lane < this->lanes; lane++){ // for lane in lanes
            for (Zombie* zombie : this->zombie_grid[lane][0]){ // for zombie in lane's 0th col
                if (zombie->entering_house == true){
                    if (this->lawnmowers[lane] == true){
                        this->lawnmowers[lane] = false;
                        kill_lane = true; // lawnmower kills all zombles in the lane
                        break;
                    }
                    else{
                        // yer ded
                        this->done = true;
                        this->win = false;
                        return;
                    }
                }
            }
            for (int col = 0; col < this->cols; col++){
                while (this->zombie_grid[lane][col].empty() == false) {
                    this->zombie_grid[lane][col].front()->get_damaged(9999, this);
                }
            }
        }
        if (this->zombies_to_spawn.empty() && this->zombie_list.empty()){
            //no  more zombies to spawn and no more alive zombies left!
            this->done = true;
            this->win = true;
        }
    }
}
bool Level::step(string plant_type, int lane, int col){
    if (lane >= this->lanes || col >= this->cols){
        return false;
    }
    this->do_zombie_actions();
    this->do_plant_actions();
    this->remove_dead_objects();
    this->do_player_action(plant_type, lane, col);
    this->spawn_zombies();
    this->spawn_suns();
    this->check_endgame();
    (this->frame)++;
    // return self.construct_state()
    return true;
}

// ====================================================================
// ====================================================================
// ====================================================================


Zombie::Zombie(int lane, int column, Level* level){
    // TODO: Placeholder values!! updaate!!!!
    lane = lane;
    col = column;
    damage = 1;
    hp = 181;
    move_interval = 10;
    attack_interval = 10;
    last_action = level->frame;
}
void Zombie::attack(Level* level){
    if((level->frame - this->last_action) < this->move_interval){
        return;
    }
    Plant* target_plant = level->plant_grid[this->lane][this->col];
    if (target_plant == nullptr) {
        return;
    }
    this->last_action = level->frame;
    target_plant->get_damaged(this->damage, level);
}
void Zombie::move(Level* level){
    if ((level->frame - this->last_action) < this->attack_interval){
        return;
    }
    this->last_action = level->frame;
    if (this->col == 0){
        level->zombie_in_home_col = true;
        this->entering_house = true;
    }
    else {
        level->zombie_grid[this->lane][this->col].remove(this);
        this->col -= 1;
        level->zombie_grid[this->lane][this->col].push_front(this);
    }
}
void Zombie::do_action(Level* level){
    this->attack(level);
    this->move(level);
    
}
void Zombie::get_damaged(int damage, Level* level){
    this->hp -= damage;
    if (this->hp > 0){
        return;
    }
    this->alive = 0;
    // remove self from both global and cell lists
    level->zombie_list.remove(this);
    level->zombie_grid[this->lane][this->col].remove(this);
    delete this;
}


// ====================================================================
// ====================================================================
// ====================================================================



Plant::Plant(int lane, int column, int frame, bool isSunflower) {
    /*
    int lane;
    int col;
    int hp;
    int damage;
    int attack_interval;
    int sun_interval;
    int last_action;
    bool isSunflower;
    */
    lane = lane;
    col = column;
    hp = 1000;
    damage = 0;
    attack_interval = 5;
    sun_interval = 5;
    last_action = frame;
    isSunflower = isSunflower;
}

Plant::Plant(int lane, int column, bool isSunflower) {
    this->prev = nullptr;
    this->next = nullptr;
    this->data = new Plant_data();
    this->data->lane = lane;
    this->data->col = column;
    this->data->hp = 300;
    if (isSunflower) {
        this->data->attack_interval = 999999;
        this->data->sun_interval = 50;
    }
    else {
        this->data->damage = 20;
        this->data->attack_interval = 10;
    }
}

void Plant::attack(Level* level) {
    this->data->last_action = level->frame;
    // pea shoot goes burr
}
void Plant::generate_sun(Level* level) {
    level->suns += 25;
}
void Plant::do_action(Level* level) {
    if (level->frame - this->data->last_action < level->frame) {
        return;
    }
    if (this->isSunflower) {
        this->generate_sun(level);
    }
    else {
        this->attack(level);
    }
}
void Plant::get_damaged(int damage, Level* level) {
    (this->data)->hp -= damage;
    if ((this->data)->hp > 0) {
        return;
    }
    Plant* next = this->next;
    Plant* prev = this->prev;
    if (next != nullptr) {
        if (prev != nullptr) {
            // prev me next
            prev->next = next;
            next->prev = prev;
        }
        else {
            // NULL me next
            next->prev = nullptr;
            level->plant_list_head = next;
        }
    }
    else { // next is null
        if (prev != nullptr){
            // prev me null
            prev->next = next;
        }
        else {
            // list empty after removal
            level->plant_list_head == nullptr;
        }
    }
}


